### PR TITLE
Add `is_zero_entry` for matrices

### LINKED ISSUE #1266: Add `is_zero_entry`
Hecke has a useful function `is_zero_entry`, and I think we should add it to AbstractAlgebra. Basic version:
```julia
is_zero_entry(M::MatrixElem, i::Int, j::Int) = iszero(M[i,j])
```
But other implementations can do more clever things. In particular, they may be able to avoid creating an auxiliary Julia object to represent the value of `M[i,j]`.

This can then be used as a building block for finding pivots (do we already have functions for finding the first/next/last non-zero entry in a row/column? If not, we should?)

### PR DESCRIPTION
**Pull request by lgoettgens**
Resolves #1266.

Specialized methods will be introduced in nemocas/Nemo.jl#1455.

### CODE REVIEW

File: src/Matrix.jl [hunk 1]
```diff
@@ -222,32 +222,41 @@ function isone(a::MatrixElem{T}) where T <: NCRingElement
...
 end
 
+@doc raw"""
+    is_zero_entry(M::MatrixElem{T}, i::Int, j::Int) where T <: NCRingElement
+
+Return `true` if $M_{i,j}$ is zero.
+"""
+@inline is_zero_entry(M::MatrixElem{T}, i::Int, j::Int) where T <: NCRingElement = iszero(M[i,j])
+
+@inline is_zero_entry(M::MatrixElem{T}, is::Tuple{Int, Int}) where T <: NCRingElement = is_zero_entry(M, is[1], is[2])
```
Comment (thofma) [inline src/Matrix.jl:236]: Is this also coming from Hecke?

Comment (lgoettgens) [reply src/Matrix.jl:236]: Hecke imports and uses the function from Nemo for `ZZMatrx`. Furthermore, it defines a similar function `_is_zero_entry` returning a boolean and the entry itself both for `ZZMatrix` and for the general case in https://github.com/thofma/Hecke.jl/blob/8c53e50b756dc2abb1bd57b6566b65d2bce747cc/src/Sparse/Matrix.jl#L274-L283. This is only used in https://github.com/thofma/Hecke.jl/blob/8c53e50b756dc2abb1bd57b6566b65d2bce747cc/src/Sparse/Matrix.jl#L316

I would suggest either removing this function from Hecke to reduce confusion or changing the interface of the proposed AbstractAlgebra and current Nemo methods to be able to remove the Hecke functions as duplicates of upstream.

**Comment by fieker** [comment]
On Mon, May 08, 2023 at 03:16:26AM -0700, Max Horn wrote:
> @fingolfin commented on this pull request.
> 
> 
> 
> >  end
>  
> ***@***.*** raw"""
> +    is_zero_entry(M::MatrixElem{T}, i::Int, j::Int) where T <: NCRingElement
> +
> +Return `true` if $M_{i,j}$ is zero.
> +"""
> ***@***.*** is_zero_entry(M::MatrixElem{T}, i::Int, j::Int) where T <: NCRingElement = iszero(M[i,j])
> 
> I don't think ***@***.***` is useful here, the compiler doesn't need a hint to inline small functions like this. But it also shouldn't hurt, so ... *shrug*
> 
> >  end
>  
> ***@***.*** raw"""
> +    is_zero_entry(M::MatrixElem{T}, i::Int, j::Int) where T <: NCRingElement
> +
> +Return `true` if $M_{i,j}$ is zero.
> +"""
> ***@***.*** is_zero_entry(M::MatrixElem{T}, i::Int, j::Int) where T <: NCRingElement = iszero(M[i,j])
> +
> ***@***.*** is_zero_entry(M::MatrixElem{T}, is::Tuple{Int, Int}) where T <: NCRingElement = is_zero_entry(M, is[1], is[2])
> 
> I don't see why we should defined an undocumented method `is_zero_entry(M::MatrixElem{T}, is::Tuple{Int, Int})` here. Unless there are strong reasons why it is needed, let's remove it. If there *are* reasons to keep it, then document it? The only scenario I can think of in the middle is "yeah we shouldn't have it, but there is some legacy code that calls it, and we can't be bothered to update it", but that wouldn't be a reason to add that method here, just one to keep that method in wherever it originally came from (Hecke.jl?). Thought ideally I'd just grep for all places using it and updating them?
> 
> I also don't understand why `_is_zero_entry` is mentioned in this
> regard, it seems independent. I've re-read the comment by @lgoettgens
> and I still don't understand. It seems like there is a separate
> discussion here, namely whether we should have a function that works
> like `_is_zero_entry` but is "official". My two cents on that: perhaps
We should not. it is either fast and efficient and dangerous
or pointless.

**Review by fingolfin — APPROVED** [review]

**Comment by lgoettgens** [comment]
@fingolfin I think the required CI jobs must be adapted from julia 1.6 to 1.8 for this to be auto-merged since the 1.6 jobs will never run.

File: src/Matrix.jl (hunk 1 above)
Comment (thofma) [reply src/Matrix.jl]: The Hecke functions do not add a method with second argument `Tuple{Int, Int}`, so I am still confused.

**Pull request merged**